js/modules/k6/data/share.go

```diff
@@ -33,103 +31,80 @@ type sharedArray struct {
 	arr []string
 }
 
-func (s sharedArray) wrap(ctxPtr *context.Context, rt *goja.Runtime) goja.Value {
-	cal, err := rt.RunString(arrayWrapperCode)
-	if err != nil {
-		common.Throw(rt, err)
-	}
-	call, _ := goja.AssertFunction(cal)
-	wrapped, err := call(goja.Undefined(), rt.ToValue(common.Bind(rt, s, ctxPtr)))
-	if err != nil {
-		common.Throw(rt, err)
-	}
+type wrappedSharedArray struct {
+	sharedArray
 
-	return wrapped
+	rt       *goja.Runtime
+	freeze   goja.Callable
+	isFrozen goja.Callable
+	parse    goja.Callable
 }
 
-func (s sharedArray) Get(index int) (interface{}, error) {
-	if index < 0 || index >= len(s.arr) {
-		return goja.Undefined(), nil
-	}
-
-	// we specifically use JSON.parse to get the json to an object inside as otherwise we won't be
-	// able to freeze it as goja doesn't let us unless it is a pure goja object and this is the
-	// easiest way to get one.
-	return s.arr[index], nil
+func (s sharedArray) wrap(rt *goja.Runtime) goja.Value {
+	freeze, _ := goja.AssertFunction(rt.GlobalObject().Get("Object").ToObject(rt).Get("freeze"))
+	isFrozen, _ := goja.AssertFunction(rt.GlobalObject().Get("Object").ToObject(rt).Get("isFrozen"))
+	parse, _ := goja.AssertFunction(rt.GlobalObject().Get("JSON").ToObject(rt).Get("parse"))
+	return rt.NewDynamicArray(wrappedSharedArray{
+		sharedArray: s,
+		rt:          rt,
+		freeze:      freeze,
+		isFrozen:    isFrozen,
+		parse:       parse,
+	})
 }
 
-func (s sharedArray) Length() int {
-	return len(s.arr)
+func (s wrappedSharedArray) Set(index int, val goja.Value) bool {
+	panic(s.rt.NewTypeError("SharedArray is immutable")) // this is specifically a type error
 }
 
-/* This implementation is commented as with it - it is harder to deepFreeze it with this implementation.
-type sharedArrayIterator struct {
-	a     *sharedArray
-	index int
+func (s wrappedSharedArray) SetLen(len int) bool {
+	panic(s.rt.NewTypeError("SharedArray is immutable")) // this is specifically a type error
 }
 
-func (sai *sharedArrayIterator) Next() (interface{}, error) {
-	if sai.index == len(sai.a.arr)-1 {
-		return map[string]bool{"done": true}, nil
+func (s wrappedSharedArray) Get(index int) goja.Value {
+	if index < 0 || index >= len(s.arr) {
+		return goja.Undefined()
 	}
-	sai.index++
-	var tmp interface{}
-	if err := json.Unmarshal(sai.a.arr[sai.index], &tmp); err != nil {
-		return goja.Undefined(), err
+	val, err := s.parse(goja.Undefined(), s.rt.ToValue(s.arr[index]))
+	if err != nil {
+		common.Throw(s.rt, err)
+	}
+	err = s.deepFreeze(s.rt, val)
```

Comment:
Could we `deepFreeze` all elements once upfront (in e.g. `getShareArrayFromCall()`), and avoid having to do this everytime `Get()` is called? I understand it's needed because it's parsed so it's a new object everytime, but if we can avoid having to stringify/parse then we could just return the same frozen object.

Reply:
see https://github.com/loadimpact/k6/pull/1848#discussion_r572818058